doc/source/data/transforming-datasets.rst

```diff
@@ -78,34 +80,338 @@ express your customized business logic in transformations. Here we will focus on
 API in Datasets.
 
-A UDF can be a function or a callable class, which has the following input/output:
+A batch UDF can be a function or, if using the actor compute strategy, a callable class.
+These UDFs have several batch format options, which control the format of the
+batches that are passed to the provided batch UDF. Depending on the underlying dataset
+format, using a particular batch format may or may not incur a data conversion cost
+(e.g. convertion an Arrow Table to a Pandas DataFrame, or creating an Arrow Table from a
+Pyhton list, both of which would incur a full copy of the data).
+
+Dataset Formats
+===============
+
+* **Tabular (Arrow or Pandas) Datasets:** Represented under-the-hood as
+  `Arrow Tables <https://arrow.apache.org/docs/python/generated/pyarrow.Table.html>`__
+  or
+  `Pandas DataFrames <https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html>`__.
+  Tabular datasets are created by reading on-disk tabular data (Parquet, CSV, JSON),
+  converting from in-memory tabular data (Pandas DataFrames, Arrow Tables,
+  dictionaries, Dask DataFrames, Modin DataFrames, Spark DataFrames, Mars DataFrames)
+  and by returning tabular data (Arrow, Pandas, dictionaries) from previous batch
+  UDFs.
+
+  Datasets will minimize conversions between the Arrow and Pandas
+  representation: tabular data read from disk will always be read as Arrow Tables, but
+  in-memory conversions of Pandas DataFrame data to a ``Dataset`` (e.g. converting from
+  Pandas, Dask, Modin, Spark, and Mars) will use Pandas DataFrames as the internal data
+  representation in order to avoid extra data conversions/copies.
+
+* **Tensor Datasets:** Represented under-the-hood as a single-column
+  `Arrow Table <https://arrow.apache.org/docs/python/generated/pyarrow.Table.html>`__
+  or
+  `Pandas DataFrame <https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html>`__,
+  using our :ref:`tensor extension type <dataset-tensor-extension-api>` to embed the
+  tensor in these tables under a single ``"__value__"`` column. Tensor datasets are
+  created by reading on-disk tensor data (NPY), converting from in-memory tensor data
+  (NumPy), and by returning tensor data (NumPy) from previous batch UDFs.
+
+  Note that converting the underlying tabular representation to a NumPy ndarray is
+  zero-copy, including converting the entire column to an ndarray and converting just
+  a single column element to an ndarray.
+
+* **Simple Datasets:** Represented under-the-hood as plain Python lists. Simple
+  datasets are created by reading on-disk binary and plain-text data, converting from
+  in-memory simple data (Python lists), and by returning simple data (Python lists)
+  from previou batch UDFs.
+
+  Simple datasets are mostly used as an escape hatch for data that's not cleanly
+  representable in Arrow Tables and Pandas DataFrames.
+
+Batch Formats
+=============
+
+Batch UDFs have the following batch format options, which control the format of the data
```

Comment:
It might be more informative to slightly adjust this into "Batch format is ...." (one-liner about what is batch format) and then the possible formats allowed.